## Modules

To use a module in a Python program it first has to be imported. 

A module can be imported using the import statement. 

For example, to import the module math, which contains many standard mathematical functions, we can do:

In [62]:
import math

In [3]:
x=math.cos(2*math.pi)
print(x)

1.0


To see all the functions in the program you just imported use the following command:

In [4]:
print(dir(math))

['__doc__', '__loader__', '__name__', '__package__', '__spec__', 'acos', 'acosh', 'asin', 'asinh', 'atan', 'atan2', 'atanh', 'ceil', 'copysign', 'cos', 'cosh', 'degrees', 'e', 'erf', 'erfc', 'exp', 'expm1', 'fabs', 'factorial', 'floor', 'fmod', 'frexp', 'fsum', 'gamma', 'gcd', 'hypot', 'inf', 'isclose', 'isfinite', 'isinf', 'isnan', 'ldexp', 'lgamma', 'log', 'log10', 'log1p', 'log2', 'modf', 'nan', 'pi', 'pow', 'radians', 'remainder', 'sin', 'sinh', 'sqrt', 'tan', 'tanh', 'tau', 'trunc']


## Variables and Types:

## Symbol Names
    Keywords that cannot be used as variable names:
   and, as, assert, break, class, continue, def, del, elif, else, except, 
   exec, finally, for, from, global, if, import, in, is, lambda, not, or,
   pass, print, raise, return, try, while, with, yield



## Assignment
    The assignment operator in Python is =. 
    
    Python is a dynamically typed language, so we do not need to specify the type of a variable when we create one.
    
    Assigning a value to a variable creates or updates the variable:


In [67]:
a=5
print(a)
print(type(a))

5
<class 'int'>


In [68]:
a=17.0
print(a)
print(type(a))

17.0
<class 'float'>


## Integers:
    int, Whole numbers, ex: 3,3000, 1

In [6]:
x=1
type(x)

int

## Floating Point:
    float, Numbers with a decimal point, ex:2.3, 4.6, 100.0

In [7]:
x=0.0
type(x)

float

## Oprerators and Comparisons
    Most operators and comparisons in Python work as one would expect:
    
    Arithmetic operators +,-,*,/,//(integer division),**(rasie to power)

This operator creates a tuple of integers

In [71]:
1 + 2, 1 - 2, 1 * 2, 1 / 2

(3, -1, 2, 0.5)

In [72]:
x=(1 + 2, 1 - 2, 1 * 2, 1 / 2)
type(x)

tuple

This operator creates a tuple of float

In [73]:
1.0 + 2.0, 1.0 - 2.0, 1.0 * 2.0, 1.0 / 2.0

(3.0, -1.0, 2.0, 0.5)

In [74]:
y=(1.0 + 2.0, 1.0 - 2.0, 1.0 * 2.0, 1.0 / 2.0)
type(y)

tuple

Note: The / operator always performs a floating point division in Python 3.x. This is not true in Python 2.x, where the result of / is always an integer if the operands are integers. to be more specific, 1/2 = 0.5 (float) in Python 3.x, and 1/2 = 0 (int) in Python 2.x (but 1.0/2 = 0.5 in Python 2.x).

In [75]:
3.0/2.0

1.5

In [79]:
1.0//1.0

1.0

** is to raise a number to a power while ^ does something weird

In [77]:
2^2

0

In [78]:
2**2

4

## Boolean operators 
    bool, Logical value indicating True or False
    
    They are spelled out as the words: and, not, or
    
    True and False must be typed as uppercase

In [80]:
True and False

False

In [81]:
not True

False

In [82]:
True or False

True

## Comparison Operators 
    x>y - x is greater than y
    
    x<y - x is less than than y
    
    x>=y - x is greater than or equal to y
    
    x<=y - x is less than or equal y
    
    x==y - x is equal to y
    x is y - x and y are identical

Whats the difference between equal and indentical? 

In [83]:
2<3

True

In [84]:
2<1

False

In [86]:
2<3,2<1

(True, False)

In [87]:
2<3,2>1

(True, True)

In [91]:
2<=2

True

In [90]:
#Equality
[1,2]==[1,2]

True

In [96]:
#Are the objects identical? 
11 = 12=[1,2]

11 is 12

#My equal signs are purple while the Johansson tutorial has 
#no highlighting for them


SyntaxError: can't assign to literal (<ipython-input-96-694ccf35c5f6>, line 2)

In [97]:
x=y=z
#Variables shouldn't be the problem
#The = sign being purple is giving me the error, I think 

NameError: name 'z' is not defined

In [98]:
11 is 12

False

In [99]:
11 is greater than 12

SyntaxError: invalid syntax (<ipython-input-99-bb63cc22da7c>, line 1)

## Compound Types: Strings, Lists,  and Dictionaries

## Strings:
    str, Ordered sequence of characters: 'hello' 'sammy' '2000'

In [8]:
x='###%%%^^^YQADGSTHS SG'
type(x)

str

In [9]:
'Aren't birds great?'

SyntaxError: invalid syntax (<ipython-input-9-3f213d0be846>, line 1)

Use double quotes when the string has apostrophies

Use single quotes when the string has double quotes

In [10]:
"Aren't Birds great?"

'He said "stolenifera"'

'He said "stolenifera"'

Only the bottom string was printed because I didn't put print before both of them. 

In [11]:
print("Aren't Birds great?")

print('He said "stolenifera"')

Aren't Birds great?
He said "stolenifera"


In [26]:
a="Hello World"
print(a)
print(type(a))
# length of the string: the number of characters, len()
print(len(a))

Hello World
<class 'str'>
11


In [28]:
#Replace a substring in a string with something else
a2=a.replace("World", "Test")
print(a2)




Hello Test


**".replace" was the code that allowed me to replace a substring with something else.**

In [29]:
print("a: " + a)
print("a2: "+ a2)

a: Hello World
a2: Hello Test


To index a character in a string, use [ ]

Indexing starts at 0

In [32]:
a[0]

'H'

# Slicing
    We can index part of a string by using the syntax [start:stop]

    This extracts the characters between index start and stop-1 

    **The character at index stop is not included.**
    
    The same methods can be used when slicing lists and strings.
    
    Returns a *slice object* representing the set of indices specified  by:
        range(start, stop, step)

In [45]:
a[0:5]

'Hello'

In [46]:
a[4:5]

'o'

If you omit either(or both) of start or stop from the [start:stop], the default is the beginning and the end of the string, respectively:

In [47]:
a[:]

'Hello World'

In [48]:
a[:6]

'Hello '

In line 45 I put a[:5] and in line 48 I put a[:6]

The difference is that line 48 includes the space after the word hello. 

In [49]:
a[5:]

' World'

We can also define the step size using the syntax [start:end:step]

The default value for step is 1

In [50]:
a[::]

'Hello World'

In [51]:
a[::1]

'Hello World'

**When the step is 1, every character is printed.**

In [53]:
a[::2]

'HloWrd'

**When the step is 2, every second character is not printed.**

In [69]:
print("str1", "str2", "str3")
#The print statement concatenates strings **with a space**
#concatenates= Link things together in a chain or series

str1 str2 str3


In [58]:
print("str1" + "str2" + "str3") 
# strings added with + are concatenated **without space**

str1str2str3


In [57]:
#The print statements converts all arguments to strings
print("str1", 1.0, False, -1j)

str1 1.0 False (-0-1j)


What do words highlighted in green mean? 

What do words highlighted in red mean? 

What do characters highlighted in purple mean?

In [59]:
print("value = %f" % 1.0)       
# we can use C-style string formatting

value = 1.000000


What is C-style string formatting?

In [60]:
# this formatting creates a string
s2 = "value1 = %.2f. value2 = %d" % (3.1415, 1.5)

print(s2)

value1 = 3.14. value2 = 1


Why is there a period before and after %2f? 

What does the % do? 

Why did the 1.5 turn into a 1? 

Why is %.2f. = 3.14 and %d=1? 

In [61]:
#alternative, more intuitive way of formatting a string 
s3 = 'value1 = {0}, value2 = {1}'.format(3.1415, 1.5)

print(s3)

value1 = 3.1415, value2 = 1.5


Why do you need to increase the value within the { } curly brackets by 1. ?

What does .format do?

**This time 1.5 stayed as 1.5 instead of turning into 1 like it did for the value of %d**

## Lists: 
    list, Ordered seqence of objects
    
    Very similar to strings, except that each element can be of any type
    
    Elements are separated by commas
    
    Lists are mutable: We can modify them by assigning new values 
    to elements in the list
   
    Ex:[10,"hello",200.3]
    
    

In [70]:
x=[10,"hello",200.3]
print(x)
print(type(x))


[10, 'hello', 200.3]
<class 'list'>


**Random Question: What is the difference between elements, characters, objects, and indexes?**

In [102]:
l = [1,2,3,4]
print(type(l))
print(l)

<class 'list'>
[1, 2, 3, 4]


The slicing function for l is nested within the print function

In [104]:
print(l[1:3])
print(l[::2])

[2, 3]
[1, 3]


How would you print the corresponding index for each element in list l?

Answer in the last lines of the notebook

In [108]:
#Python lists can be arbitrarily nested
#Here, 5 is nested in 4
#4 is nested in 3
#3 is nested in 2
#2 is nested in 1
#5,4,3,and 2 are nested in 1
nested_list=[1,[2,[3,[4,[5]]]]]
nested_list

[1, [2, [3, [4, [5]]]]]

**There are a number of convenient functions for generating lists of various types**

Example: the range function

In [109]:
start=10
stop=30
step=2

range(start, stop, step)

range(10, 30, 2)

Cell 190 didn't give me what I wanted.

I wanted the number assigned to variables start,stop,and step to come out in the function range()

In [110]:
start=10
stop=30
step=2

In [111]:
range(start, stop, step)

range(10, 30, 2)

Separating the variable assignment and the range function in two different cells didn't work. 

In [112]:
range(10,30,2)

range(10, 30, 2)

It still didn't work. There must be a problem with my range function then.

In [114]:
range[10:30:2]

TypeError: 'type' object is not subscriptable

What is an iterator? 

In [115]:
list(range(-10,10))

[-10, -9, -8, -7, -6, -5, -4, -3, -2, -1, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [118]:
list(range(10,30))

[10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 24,
 25,
 26,
 27,
 28,
 29]

In [119]:
list(range(10,30,2))

[10, 12, 14, 16, 18, 20, 22, 24, 26, 28]

So now it worked! But, you must nest the range function inside the list function

In [120]:
print(range(10,30,2))

range(10, 30, 2)


Print doesn't work the same and the list function for nesting the range function

In [121]:
s

'Hello World'

In [123]:
#Convert a string to a list by type casting:
s2=list(s)
s2

['H', 'e', 'l', 'l', 'o', ' ', 'W', 'o', 'r', 'l', 'd']

In [125]:
#Sorting a list
s2.sort()
s2

[' ', 'H', 'W', 'd', 'e', 'l', 'l', 'l', 'o', 'o', 'r']

In what order did it sort the list assigned to variable s2?

It is alphabetical in the tutorial but not for me.

## Adding, inserting, modifying, and removing elements from lists

In [126]:
#Create a new empty list
l=[]

In [127]:
#Add an element using 'append'
l.append("A")
l.append("d")
l.append("d")

print(l)

['A', 'd', 'd']


In [128]:
#Modify list by assigning new values to elements in the list
l[1]='p'
l[2]='p'

l

['A', 'p', 'p']

In [129]:
#Insert an element at an specific index using insert

l.insert(0, "i")
l.insert(1, "n")
l.insert(2, "s")
l.insert(3, "e")
l.insert(4, "r")
l.insert(5, "t")

print(l)


['i', 'n', 's', 'e', 'r', 't', 'A', 'p', 'p']


In [137]:
print(l[-2:])

['p', 'p']


In [139]:
l[-2:]=["d","d"]
l

['i', 'n', 's', 'e', 'r', 't', 'A', 'd', 'd']

**Remove first element with specific value using 'remove'**


In [ ]:
l.remove("A")

In [143]:
l

['i', 'n', 's', 'e', 'r', 't', 'd', 'd']

In [148]:
l.remove('d')
l

['i', 's', 'e', 'r', 't', 'd']

Note: .remove('d') only removed the first d

**Remove an element at a specific location using del:**


In [145]:
del l[1]


In [146]:
l

['i', 's', 'e', 'r', 't', 'd', 'd']

# Tuples
    tup, Ordered objects like lists, except they have ( ) parenthesis instead of [ ] brackets or they are just separated by commas
    
    Immutable: they cannot be modified once they are created
    
    Ex: 1,2,3 or (1,2,3)

In [150]:
x= 1,2,3
print(type(x))

<class 'tuple'>


In [151]:
x= 1,2,3
x

(1, 2, 3)

In [152]:
x= 1,2,3
type(x)

tuple

We can unpack a tuple by assigning it to a comma-separated list of variables:


In [154]:
point= 10, 20

In [155]:
x,y = point

print("x =", x)
print("y =", y)

x = 10
y = 20


In [157]:
point[0]=20

TypeError: 'tuple' object does not support item assignment

I got an Error for trying to assign a new value to an element in a tuple

# Dictionaries
    dict, Unordered Key: Value pairs:
    
    Like lists, except that each element is a key-value pair. 
        
    Ex: {"mykey':"value", "name":Frankie"}
    
    The syntax for dictionaries is {key1:value1, ...}:

In [22]:
params = {"parameter1" : 1.0,
          "parameter2" : 2.0,
          "parameter3" : 3.0,}

print(type(params))
print(params)

<class 'dict'>
{'parameter1': 1.0, 'parameter2': 2.0, 'parameter3': 3.0}


**Random Question: Whats the difference between type and class**

In [23]:
print("parameter1 = " + str(params["parameter1"]))
print("parameter2 = " + str(params["parameter2"]))
print("parameter3 = " + str(params["parameter3"]))



parameter1 = 1.0
parameter2 = 2.0
parameter3 = 3.0


In [24]:
#change parameter 1 and 2
params["parameter1"] = "A"
params["parameter2"] = "B"

# add a new entry
params["parameter4"] = "D"

print("parameter1 = " + str(params["parameter1"]))
print("parameter2 = " + str(params["parameter2"]))
print("parameter3 = " + str(params["parameter3"]))
print("parameter4 = " + str(params["parameter4"]))

parameter1 = A
parameter2 = B
parameter3 = 3.0
parameter4 = D


Sets: set, unordered collection of unique objects: {"a","b"}

# Type Casting 

In [25]:
x=1.5
print(x,type(x))

1.5 <class 'float'>


In [26]:
x=int(x)
print(x,type(x))

1 <class 'int'>


In [27]:
z=complex(x)
print(z,type(z))

(1+0j) <class 'complex'>


In [28]:
x=float(z)

TypeError: can't convert complex to float

Complex numbers cannot be casted to float or int

In [ ]:
y=bool(z.real)
y

In [ ]:
print(z.real, "-->", y, type(y))

How can a number be true?

In [ ]:
y=bool(z.imag)
print(y)
print(z.imag,"-->", y, type(y))


In [ ]:
z

In [ ]:
bool(z)

You cant switch complex numbers to floats or integers, but you can switch them to bool. 

I don't understand the result of casting a complex number to bool

# Control Flow

**Conditional Statements: if, elif, else**



The Python syntax for conditional execution of code uses the keywords:
- if
- elif (else if)
- else:


In [ ]:
statement1 = False
statement2 = False

if statement1:
    print("statement1 is True")
    
elif statement2:
    print("statement2 is True")
    
else:
    print("statement1 and statement2 are False")


In [ ]:
statement1 = statement2 = False

if statement1:
    if statement2:
        print("both statement1 and statement2 are True")
        
#Nothing is Printed since nothing was specified for false statements

In [ ]:
statement1 = True 

if statement1:
    print("printed if statement1 is True")
    
    print("still inside the if block")

In [ ]:
statement1 = False
if statement1:
    print("printed if statement1 is True")
    
print("now outside the if block")

#'Now outside the block' prints regardless of whether its true or false

# Loops

In python, loops can be programmed in a number of different ways. the most common is the for loop

It is used together with iterable objects, such as lists.


In [ ]:
for x in [1,2,3]:
    print(x)
#x=each element. []=list

The for loop iterates over the elements of the supplied list, and executes the containing block once for each element. Any kind of list can be used in the for loop. For example:

In [ ]:
for x in range(4): # by default range start at 0
    print(x)

In [ ]:
for x in range(-3,3):
    print(x)

In [ ]:
for word in ["scientific", "computing", "with", "python"]:
    print(word)

In [ ]:
parameter1 = 'A'
parameter2 = 'B'
parameter3 = 3.0
parameter4 = 'D'

In [29]:
for key, value in params.items():
    print(key + " = " + str(value))

parameter1 = A
parameter2 = B
parameter3 = 3.0
parameter4 = D


In [31]:
for idx, x in enumerate(range(-3,3)):
    print(idx, x)
#x= the index
#y= the element in the range

0 -3
1 -2
2 -1
3 0
4 1
5 2


List comprehensions: Creating lists using for loops:

In [32]:
#Note: function is placed before the loop. 
list1 = [x**2 for x in range(0,5)]

print(list1)

[0, 1, 4, 9, 16]


**While loops:**

In [33]:
i = 0

while i < 5:
    print(i)
    
    i = i + 1
    
print("done")

0
1
2
3
4
done


In [40]:
i = 10

while i<20:
    print(i)
    
    i = i + 1

10
11
12
13
14
15
16
17
18
19
